stdenv/check-meta: deep type checks

ncfavier opened the pull request:
Check `meta` attributes deeply: `meta.maintainers = [ 42 ];` now results in an evaluation error if `checkMeta` is enabled. Adds a `lib.typeCheck` function that performs deep type checks.

Prior context (optional): https://github.com/NixOS/nixpkgs/pull/6794#issuecomment-1003397767, https://github.com/NixOS/nixpkgs/pull/191171, https://github.com/NixOS/nixos-search/issues/391

**Note: this does NOT set `checkMeta` to `true` by default, it only changes what happens when it is.**

I used the following command (extracted from OfBorg) to run the checks:

```sh
nix-env --json -f '<nixpkgs>' -I nixpkgs=. --arg config '{
  checkMeta = true;
  handleEvalIssue = reason: msg:
    if builtins.elem reason ["unknown-meta" "broken-outputs"] then abort msg else true;
}' -qa --out-path --meta > /dev/null
```
EDIT: this is not enough as it does not recurse into attrsets, I'd have to use OfBorg's [outpaths.nix](https://github.com/NixOS/ofborg/blob/released/ofborg/src/outpaths.nix) but my computer doesn't have enough memory...

It would be nice to abstract this into a script somewhere in nixpkgs, so that people can run it locally without needing OfBorg (and OfBorg can then just call it). Where should it go? `maintainers/scripts`? `nixpkgs-basic-release-checks.nix`?

I've left a few TODOs in the code, they don't have to be addressed in this PR (but they could).

Ping everyone remotely involved, feel free to ignore if you're not interested: @piegamesde @Profpatsch @sternenseemann @roberth @tazjin @SuperSandro2000 @7c6f434c @mweinelt @amjoseph-nixpkgs @K900

ncfavier committed:
treewide: fix broken `meta` attributes

ncfavier committed:
lib/customisation.overrideDerivation: propagate evaluation condition

The new derivation should evaluate only if the old derivation does.

Sadly this means that the old derivation cannot depend on the new one
any more, which was used by xorgserver on Darwin. But this is not a
problem as `overrideAttrs` can (and should) usually be used instead.

This change allowed catching an invalid `meta.platforms` in the linux_rpi
kernels, which use `overrideDerivation`.

roberth commented on lib/modules.nix:
```diff
@@ -772,6 +772,17 @@ rec {
+      { file = unknownModule; inherit value; }
+    ];
+    eval = builtins.tryEval (builtins.deepSeq merged.mergedValue null);
+  in eval.success;
```
The part before `tryEval` and `deepSeq` could be called `typeChecking`: to return the same value, which is lazily checked on demand. Unlike `fix (a: typeCheck t (f a))`, `typeChecking` would allow `f` to use `a` in parts of its return value.
One alternative naming scheme for these two functions would be `strictCheck` / `lazyCheck`.

ncfavier replied:
Sounds good, I was thinking of adding something like that for situations where we want an error message rather than `false`. The only question is whether `typeChecking` should take a file argument or use `unknownModule`.

roberth commented on lib/modules.nix:
```diff
@@ -772,6 +772,23 @@ rec {
+     of the module system.
+     Note that calling `type.check` is not enough: see `mkOptionType`'s documentation.
+  */
+  typeCheck = type: value: let
+    eval = builtins.tryEval (builtins.deepSeq (typeChecking { inherit type value; }) null);
```
I guess it doesn't add much value, but you could simplify this to `typeChecking args`.

It improves `pos`, but that only shows up in a hypothetical `abort` message, but that should be a `throw` anyway. So the values is only in simplification and uniformity.

ncfavier replied:
I think I prefer the simplicity of `typeCheck t v`.

roberth replied:
Uniformity isn't as strong as interchangeability anyway, and we'll want to curry this one, so fair enough.

roberth commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -310,16 +303,16 @@ let
     else
-      "key 'meta.${k}' is unrecognized; expected one of: \n\t      [${lib.concatMapStringsSep ", " (x: "'${x}'") (lib.attrNames metaTypes)}]";
+      "key 'meta.${k}' is unrecognized; expected one of: \n  [${lib.concatMapStringsSep ", " (x: "'${x}'") (lib.attrNames metaTypes)}]";
   checkMeta = meta: if config.checkMeta then lib.remove null (lib.mapAttrsToList checkMetaAttr meta) else [];
```
(fyi) Conceptually, this could be `typeCheck` on a submodule, but that would allow module syntax to be used, which is not ok.

ncfavier commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -247,16 +247,7 @@ let
-  # purposes and in the hope that they will be used eventually.
-  # See https://github.com/NixOS/nixpkgs/pull/191171 for an attempt
-  # to fix this, or mkOptionType in lib/types.nix for more information.
+  # TODO make this into a proper module and use the generic option documentation generation?
```
Wouldn't disallowing raw module syntax be as easy as `{ ... }: { config = meta; }` ?

Somewhat tangentially, the module system is already leaking in an undesirable way:

```
nix-repl> lib.typeCheck lib.types.str (lib.mkIf true "a")
true
```

This is hard to fix because `type.check` and `mergeDefinitions` are so tightly coupled.

roberth replied:
You're right. We should be checking type outputs, not type inputs, and there's not type method for that yet.

roberth commented on lib/modules.nix:
```diff
@@ -772,6 +772,23 @@ rec {
+     error if the type doesn't match.
+  */
+  typeChecking = { type, value, pos ? builtins.unsafeGetAttrPos "value" args, loc ? [ "value" ] }@args: let
+    merged = mergeDefinitions loc type [
+      { file = pos.file or unknownModule; inherit value; }
+    ];
+  in merged.mergedValue;
```
The names of types describe the output type, and it would be reasonable to expect this to check what the name of the type suggests.
However, as you know, this checks module system definition inputs, and not type outputs. This is not an acceptable implementation; at least not for `lib`.
If you don't want to implement an output checking method on the types, you may choose to move these functions to check-meta, with an explanatory comment.

ncfavier replied:
I'm not entirely sure what you mean by type inputs and outputs. Is this about my earlier comment about `mkIf`? I don't really see a solution to that that doesn't involve a deep refactoring of the module system.

roberth replied:
Yes, `mkIf` is one of the differences between type inputs and outputs.
They're not really types but functions, the way they're part of the module system. We try to write idempotent types, but that's not a guarantee.

> deep refactoring

A simple change would be to add a new method to the types; to consider output checking as a completely new use case for types that's distinct from types' use in the module system.
I don't think the module system's merging and checking should be decoupled, because it's better to check the inputs, for better error messages, and it's better not to check the output and rely on a correct merge implementation, because otherwise, we're checking things twice (or perhaps even quadratically), which is slow for no real benefit.

ncfavier replied:
So you're suggesting something like `checkOutput = v: isList v && all elemType.checkOutput v`, for `listOf elemType`? That seems like a lot of code duplication, but I guess we could at least make `checkOutput` default to `check` if not given.

In any case I'll probably just move the functions out of `lib` for this PR.

roberth replied:
I guess you could call that duplication, yes. They're similar code paths to achieve similar goals, but there's no opportunity for code reuse a the type interface level, and any reuse achieved by a `mkOptionType` helper function would be disproportionally slow, considering how widely this code is going to be called.
So yeah, the implementation is not pretty, but what matters most is how it behaves, and the interface towards users of types and the module system.



> In any case I'll probably just move the functions out of `lib` for this PR.

sgtm



> something like `checkOutput = v: isList v && all elemType.checkOutput v`,

Maybe. This does not allow lazy checking, or nice error messages that can be listed, like the assertions and warnings in NixOS. Can be discussed later.

ncfavier committed:
stdenv/check-meta: do deep type checks

Use a wrapper around `mergeDefinitions` to type-check values deeply, so
that e.g. `maintainers = [ 42 ];` is an error.
